aesd-char-driver/main.c

```diff
@@ -0,0 +1,241 @@
+#include <linux/cdev.h>
+#include <linux/string.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Ganesh KM"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    PDEBUG("open");
+    /**
+     * TODO: handle open
+     */
+    struct aesd_dev *dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+	filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+		
+	struct aesd_dev* device = filp->private_data;
+	struct aesd_buffer_entry * read_entry = NULL;
+	size_t bytes_to_read = 0;
+	size_t entry_offset_byte = 0;
+	size_t available_bytes = 0;
+	ssize_t retval = 0;	
+	
+	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+	/**
+	 * TODO: handle read
+	 */
+	 
+	if(mutex_lock_interruptible(&device->char_device_lock))
+		return -ERESTARTSYS; 
+	
+	read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&device->aesd_buffer, *f_pos, &entry_offset_byte);	
+		
+	if(read_entry == NULL) 
+	{
+	   retval = 0;
+	   goto out;
+	} 
+	
+	available_bytes = read_entry->size - entry_offset_byte;
+	
+	if(available_bytes > count)
+	{
+	   bytes_to_read = count;
+	}
+	else
+	{ 
+	   bytes_to_read = available_bytes;
+	}
+	
+	int ret = copy_to_user(buf, read_entry->buffptr + entry_offset_byte, bytes_to_read);
+
+	if(ret)
+	{
+	   retval = -EFAULT;
+	   goto out;
+	}
+	
+	retval= bytes_to_read;
+	
+	*f_pos += bytes_to_read;
+	
+out: 
+	mutex_unlock(&device->char_device_lock);	
+	return retval;	
+	 
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+	ssize_t retval = -ENOMEM;
+	PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+	/**
+	 * TODO: handle write
+	 */
+	 
+	const char* ret_entry = NULL;
+	struct aesd_dev* device = filp->private_data;
+	
+	if(mutex_lock_interruptible(&device->char_device_lock))
+		return -ERESTARTSYS;
+		
+	if(device->add_entry.size == 0)
+	{
+	   device->add_entry.buffptr = kzalloc(count, GFP_KERNEL);
+	}	
+	else 
+	{
+	   device->add_entry.buffptr = krealloc(device->add_entry.buffptr, device->add_entry.size + count, GFP_KERNEL);
+	}
+	 
+	if(device->add_entry.buffptr == NULL)
+	{
+	   retval = -ENOMEM;
+	   goto out;
+	}  
+	 
+	retval = count;
+	
+	size_t error_bytes_num = copy_from_user((void*)(&device->add_entry.buffptr[device->add_entry.size]), buf, count);
+	
+	if(error_bytes_num)
+	{
+	   retval = retval - error_bytes_num;
+	}
+	
+	device->add_entry.size += retval;
+	
+	
+	if(strchr((char*)(device->add_entry.buffptr), '\n'))
```

Comment:
@Ganeshkm10 use `strnchr` here to handle the case where you were passed a buffer which wasn't null terminated.